include/ofi_mb.h

```diff
@@ -0,0 +1,204 @@
+ * to redistribute to others.  This software was authored by:
+ *
+ * Mathematics and Computer Science Division
+ * Argonne National Laboratory, Argonne IL 60439
+ *
+ * (and)
+ *
+ * Department of Computer Science
+ * University of Illinois at Urbana-Champaign
+ *
+ *
+ * 			      GOVERNMENT LICENSE
+ *
+ * Portions of this material resulted from work developed under a U.S.
+ * Government Contract and are subject to the following license: the Government
+ * is granted for itself and others acting on its behalf a paid-up,
+ * nonexclusive, irrevocable worldwide license in this computer software to
+ * reproduce, prepare derivative works, and perform publicly and display
+ * publicly.
+ *
+ * 				  DISCLAIMER
+ *
+ * This computer code material was prepared, in part, as an account of work
+ * sponsored by an agency of the United States Government.  Neither the United
+ * States, nor the University of Chicago, nor any of their employees, makes any
+ * warranty express or implied, or assumes any legal liability or responsibility
+ * for the accuracy, completeness, or usefulness of any information, apparatus,
+ * product, or process disclosed, or represents that its use would not infringe
+ * privately owned rights.
+ */
+/*
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "config.h"
+#include <stdbool.h>
+
+#ifdef HAVE_ATOMICS
+#include <stdatomic.h>
+
+static inline void ofi_wmb(void)
+{
+	atomic_thread_fence(memory_order_release);
+}
+
+#elif defined(HAVE_BUILTIN_MM_ATOMICS)
+
+static inline void ofi_wmb(void)
+{
+	__atomic_thread_fence(__ATOMIC_RELEASE);
+}
+
+#else
+#error "No atomics support found for SHM provider."
```

Comment:
Actually just noticed this shm-specific error code here and no fallback. That's going to fail non atomics builds. Please add a fallback or else remove the error else but add a requirement of HAVE_ATOMICS on shm

Reply:
> That's going to fail non atomics builds. Please add a fallback or else remove the error else but add a requirement of HAVE_ATOMICS on shm

Fallback would be hard to implement, because it would require us to write assembly code for every CPU architechture.

I have updated the error message, also updated shm's configure.m4 to check HAVE_ATOMICS

Reply:
@aingerson  I think I addressed your comments, please take a look. :)